src/Razor/src/Microsoft.AspNetCore.Razor.VSCode/src/RazorCodeLensProvider.ts

```diff
@@ -0,0 +1,130 @@
+                    remappedCodeLenses.push(newCodeLens);
+                } else {
+                    // This means this CodeLens was for non-user code. We can safely ignore those.
+                }
+            }
+
+            return remappedCodeLenses;
+
+        } catch (error) {
+            return [];
+        }
+    }
+
+    public async resolveCodeLens(codeLens: vscode.CodeLens, token: vscode.CancellationToken) {
+        if (codeLens instanceof RazorCodeLens) {
+            return await this.resolveRazorCodeLens(codeLens, token);
+        }
+    }
+
+    private async resolveRazorCodeLens(codeLens: RazorCodeLens, token: CancellationToken): Promise<vscode.CodeLens> {
+        // Initialize with default values.
+        codeLens.command = {
+            title: '',
+            command: '',
+            arguments: []
+        };
+
+        try {
+            const razorDocument = await this.documentManager.getDocument(codeLens.uri);
+            if (!razorDocument) {
+                return codeLens;
+            }
+
+            // Make sure this CodeLens is for a valid location in the projected C# document.
+            const projection = await this.getProjection(codeLens.document, codeLens.range.start, token);
+            if (!projection || projection.languageKind != LanguageKind.CSharp) {
+                return codeLens;
+            }
+
+            const references = await vscode.commands.executeCommand<vscode.Location[]>(
+                'vscode.executeReferenceProvider',
+                projection.uri,
+                projection.position) as vscode.Location[];
+
+            // We now have a list of references to show in the CodeLens.
+            // Let's remap any locations pointing to any background documents back to the original documents.
+            // Technically, Omnisharp should have already remapped those. We're just being extra cautious here.
+            const remappedLocations = new Array<vscode.Location>();
+            for (const reference of references) {
+                if (!isRazorCSharpFile(reference.uri)) {
+                    // This is a regular C# file. No need to re-map.
+                    remappedLocations.push(reference);
+                    continue;
+                }
+
+                const razorFile = getRazorDocumentUri(reference.uri);
+
+                // Re-map the projected reference range to the host document range
+                const result = await this.serviceClient.mapToDocumentRange(
+                    projection.languageKind,
+                    reference.range,
+                    razorFile);
+
+                if (result) {
+                    remappedLocations.push(new vscode.Location(razorDocument.uri, result.range));
+                }
+            }
+
+            const count = remappedLocations.length;
+            codeLens.command = {
+                title: count === 1 ? '1 reference' : `${count} references`,
```

Comment:
Hollllly, I had no idea this is how the CodeLens actions worked. Dope dope dope 👏 🥂 🎉